test/select_io_dset.c

```diff
@@ -0,0 +1,3212 @@
+                            for (j = 0; j < DSET_SELECT_DIM; j++)
+                                if (s3_rbufi3[i][j].a != s1_wbufi3[i][j].a ||
+                                    s3_rbufi3[i][j].b != s1_wbufi3[i][j].b ||
+                                    s3_rbufi3[i][j].c != s1_wbufi3[i][j].c ||
+                                    s3_rbufi3[i][j].d != s1_wbufi3[i][j].d) {
+                                    H5_FAILED();
+                                    HDprintf("    Read different values than written.\n");
+                                    HDprintf("    For dset %d at index %d\n", i, j);
+                                }
+                        }
+                        else if (s == SETTING_B) {
+                            for (j = 0; j < DSET_SELECT_DIM; j++)
+                                if (s1_rbufi3[i][j].a != s1_wbufi3[i][j].a ||
+                                    s1_rbufi3[i][j].b != (DSET_SELECT_DIM + j) ||
+                                    s1_rbufi3[i][j].c != s1_wbufi3[i][j].c ||
+                                    s1_rbufi3[i][j].d != (DSET_SELECT_DIM + j)) {
+                                    H5_FAILED();
+                                    HDprintf("    Read different values than written.\n");
+                                    HDprintf("    For dset %d at index %d\n", i, j);
+                                }
+                        }
+
+                        break;
+
+                    case DSET_NTTYPES:
+                    default:
+                        TEST_ERROR;
+
+                } /* end switch dset_types */
+
+            } /* end for i ndsets */
+
+        } /* end for s settings */
+
+        /* Closing */
+        if (H5Pclose(dcpl) < 0)
+            FAIL_STACK_ERROR;
+        if (H5Pclose(dxpl) < 0)
+            FAIL_STACK_ERROR;
+
+        if (H5Tclose(s1_tid) < 0)
+            FAIL_STACK_ERROR;
+        if (H5Tclose(s3_tid) < 0)
+            FAIL_STACK_ERROR;
+        if (H5Tclose(s4_tid) < 0)
+            FAIL_STACK_ERROR;
+
+        for (i = 0; i < (int)ndsets; i++) {
+            if (H5Sclose(file_sids[i]) < 0)
+                FAIL_STACK_ERROR;
+            if (H5Dclose(dset_dids[i]) < 0)
+                FAIL_STACK_ERROR;
+            /* Don't delete the last set of datasets */
+            if ((n + 1) != niter)
+                if (H5Ldelete(fid, dset_names[i], H5P_DEFAULT) < 0)
+                    FAIL_STACK_ERROR;
+        }
+
+        /* Freeing */
+        HDfree(total_wbuf1);
+        HDfree(total_wbuf1_bak);
+        HDfree(total_rbuf1);
+
+        HDfree(ul_total_wbuf2);
+        HDfree(ul_total_wbuf2_bak);
+        HDfree(l_total_rbuf2);
+        HDfree(l_total_wbuf2);
+        HDfree(l_total_wbuf2_bak);
+        HDfree(s_total_rbuf2);
+
+        HDfree(s1_total_wbuf3);
+        HDfree(s1_total_wbuf3_bak);
+        HDfree(s3_total_rbuf3);
+        HDfree(s4_total_wbuf3);
+        HDfree(s4_total_wbuf3_bak);
+        HDfree(s1_total_rbuf3);
```

Comment:
All these buffers above should probably be set to NULL. Otherwise if the next iteration fails somewhere we might double-free segfault in the error cleanup section.